### PR TITLE
feat(adapters): annotate present-but-empty tool outputs (DeepSeek default)

### PR DESCRIPTION
**Pull request by harryzhou2000**
## Summary

- Empty tool results (a script ran without calling `text(...)`) were forwarded as an empty string, so routed models silently accepted them or re-issued the same call.
- Adds an opt-in provider option `annotateEmptyToolOutputs` that rewrites present-but-empty tool outputs to an explicit `[ocx] empty tool output ...` annotation on the OpenAI-compatible Chat and Responses wires; non-empty outputs and missing-result placeholders stay byte-identical.
- Seeds the option `true` for the DeepSeek registry entry and backfills existing DeepSeek configs; every other provider keeps legacy behavior unless the option is set explicitly.
- Management boundary: POST/PATCH validate the boolean, PATCH can set or clear it with `null`, and provider-name errors are redacted.

**Upgrade note:** existing DeepSeek configs start annotating empty tool outputs on the next upgrade; set `annotateEmptyToolOutputs: false` in the provider row to keep the legacy empty-string behavior.

## Test plan

- `bun test tests/empty-tool-output-annotation.test.ts tests/management-provider-validation.test.ts tests/responses-stateless-dangling-call-repair.test.ts` — 122 pass / 0 fail
- `bun run typecheck` — clean; `git diff --check` — clean

## Verification

- Rebased on latest `upstream/dev` (`223a0a287`) before push; head `182d881db`

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [x] All CI tests are green on my local testing.
- [x] I pushed my PR to the latest dev commit.
- [x] I resolved all correct Codex and CodeRabbit findings.

- [x] My PR is ready for review.

### CODE REVIEW

**Comment by harryzhou2000** [comment]
Addressed the CodeRabbit finding: `toolResultTextForWire` now annotates whitespace-only text-part arrays when `annotateEmptyToolOutputs` is enabled (the Responses adapter already treated them as empty), while preserving normal text and image handling. Added regression tests for the whitespace-only array (enabled and absent cases) and an image-with-whitespace case. Pushed as 7e6e986a7.

**Comment by harryzhou2000** [comment]
Added the requested orphaned-result regression tests (2d3ddf63f): an empty tool result without a matching call is annotated when `annotateEmptyToolOutputs` is enabled and stays unchanged when absent. 25 tests pass in the chat suite; typecheck clean.

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 52 / 80

설명: 이 PR은 도구가 실행됐는데 stdout 이 비어 있는 결과를 빈 문자열로 그대로 보내면, 라우트된 모델이 성공으로 받아들이거나 같은 호출을 다시 하는 구멍을 막는다. 지금 CURRENT `dev` HEAD는 `ced9a85c5` 이다. 이 시간에 착지한 것은 문서뿐이다: `#2348` WP4, `#2349` GO. 마지막 코드는 `#2346` Opus Fast. 현재 HEAD 의 openai-chat 은 toolResultTextForWire 가 문자열이면 그대로, 배열이면 텍스트만 이어 붙인다. 빈 문자열과 빈 배열을 모델이 보게 바꾸는 스위치는 없다. 이 PR은 provider 옵션 annotateEmptyToolOutputs 를 넣고, 켜져 있으면 `[ocx] empty tool output ...` 문구로 바꾼다. DeepSeek 레지스트리에만 true 를 시드하고, 기존 DeepSeek 설정에도 backfill 한다. 다른 프로바이더는 명시하기 전엔 예전 동작이다. Chat 과 Responses 둘 다 손본다. 화이트스페이스-only 텍스트 배열은 CodeRabbit 지적 후 Chat 쪽도 Responses 와 같이 빈 것으로 본다. 드래프트이고 본문 체크리스트는 0/4 다. 파일에 `src/types/provider.ts` 와 `src/config.ts` 스키마가 같이 들어 있다. types.ts/config.ts 스플릿이 진행 중이면 이 PR이 무효화되기 쉽다. 스플릿이 먼저 착지하면 리베이스하지 말고 닫고 다시 연다. Cursor `#2334` 는 여전히 cursor-pool 모듈+테스트만. `#2332` H2 는 discovery 전용. `#2320` overflow + `#2342` size prior 는 `dev`. 카탈로그 팁은 Ox Alpha `x-preview-f-free` + `deepseek-v4-flash-vision-exp`. package.json 2.27.0, 태그 v2.29.0. `#2188` 사이드카는 이미 `dev`. DeepSeek 빈 도구 결과가 조용히 통과하는 정합성 구멍이라 52. 다만 기본값을 기존 설정에 심는 일과 드래프트 게이트가 점수를 낮춘다.

src/types/provider.ts annotateEmptyToolOutputs 주석 - "the same call" 다음에 마침표가 없고 바로 Non-empty 가 이어진다
src/providers/registry.ts DeepSeek annotateEmptyToolOutputs true - 기존 설정에도 backfill. 빈 결과를 성공으로 보던 동작이 업그레이드만으로 바뀐다
src/adapters/openai-chat.ts toolResultTextForWire - 이미지 파트가 섞인 배열은 every(text) 가 실패해서 화이트스페이스-only 로 안 본다. 이미지+빈텍스트는 예전대로다
src/config.ts providerConfigSchema / src/types/provider.ts - 스키마와 타입이 한 PR. 스플릿이 먼저 착지하면 이 PR은 닫고 다시 연다
PR 체크리스트 0/4 드래프트 - 로컬 CI/최신 dev/봇 지적/레디 표시가 아직이다. 지금 머지 대상이 아니다

메인테이너의 판단이 필요한 지점
- DeepSeek 기본 true backfill 을 유지할지, 신규 설정만 시드하고 기존 사용자는 옵트인할지
- 주석 문구 `[ocx] empty tool output ...` 를 이 영어로 고정할지
- `#2351`/`#2355` 도 `src/config.ts` 를 만진다. 세 PR을 어떤 순서로 둘지

너의 추천
드래프트를 유지한다. 체크리스트 4칸을 채운 뒤에만 레디. DeepSeek 기본값 backfill 은 본문대로 두되, 업그레이드 노트를 한 줄 적는다. 주석 마침표만 고친다. `#2351` 감사 로그, `#2355` divergence 경고와 한 장에 섞지 않는다. config.ts 충돌이 나면 이 스키마 한 줄짜리부터 다시 연다. types.ts/config.ts 스플릿이 provider 필드를 이미 옮긴 뒤에야 충돌이 보이면 리베이스하지 말고 닫고 다시 연다. 지금은 그 정도 아님. 라벨은 그대로 둔다. 프리뷰 배포가 아니다.

이 댓글은 grok-bot이 작성했습니다

**Comment by harryzhou2000** [comment]
Addressed the maintainer review: ended the comment sentence in `src/types/provider.ts` (b1b5b071e) and added the one-line upgrade note to the description. DeepSeek backfill stays as reviewed; image+whitespace arrays keep legacy handling. All gates green and the PR is ready.

**Comment by lidge-jun** [comment]
## Blocker: the Responses emptiness check wipes real image and encrypted tool outputs

The Chat half of this PR is right, and the earlier CodeRabbit and maintainer items are genuinely closed at `b1b5b071`. One defect remains, and it corrupts real payloads rather than just annotating empty ones.

`isToolOutputEmpty` in `src/adapters/openai-responses.ts` treats **any part without non-empty `text`/`refusal` as empty**:

```ts
return output.every(part => {
  if (!isPlainObject(part)) return true;
  if (typeof part.text === "string" && part.text.trim() !== "") return false;
  if (part.type === "refusal" && typeof part.refusal === "string" && part.refusal.trim() !== "") return false;
  return true;   // <- input_image, encrypted_content, input_file all land here
});
```

So an `input_image`, an `encrypted_content` blob, or a `file_id`-only image is classified empty and **replaced** with the `[ocx]` annotation. The model never sees it.

Your own Chat implementation gets this right and shows the intended rule:

```ts
if (annotateEmpty && content.every(part => part.type === "text") && text.trim() === "") {
```

That refuses to annotate a mixed array. Responses has no equivalent guard.

This matters because the flag is **on by default for DeepSeek** (seeded in `registry.ts`, backfilled in `derive.ts` / `router.ts`), and DeepSeek V4 flash/pro go out over the Responses wire. Existing configs pick it up on upgrade unless they explicitly set `false`. A live probe at your head confirmed image-only, image+whitespace, `encrypted_content`-only, and `file_id`-only payloads all come back as the annotation.

### Why CI stayed green

The Responses tests only cover `""`, `"   "`, and `"ok"` — strings. They never send the content-part array Codex actually produces. Reverting the adapter hunks fails 7 tests (so the string path is genuinely covered), but the corruption path is unrepresented, which is exactly how a green suite ships this.

### To unblock

1. Treat `input_image`, `encrypted_content`, `input_file`, and any non-text part as **non-empty**. Mirror the Chat rule: annotate an array only when every part is text/refusal and the joined text trims empty.
2. Add Responses regressions for image-only, image+whitespace, `encrypted_content`, and `file_id`-only. A test that still passes after reverting `isToolOutputEmpty` is not covering this.

### Two smaller notes

- `annotateEmptyToolOutputs` is in the `config.ts` schema but the management write path in `src/server/auth-cors.ts` does not type-check it, so `POST`ing `"yes"` is accepted and only fails later at load. `responsesSnapshotRepair` has the boolean guard to copy. (`statelessResponses` and `requiresAdjacentResponsesToolResults` share the gap — not yours to fix here.)
- The 157-character annotation string is duplicated in both adapters. One shared helper would keep the two emptiness contracts from drifting again.

Leaving open rather than closing — the feature is worth having and the Chat side is already correct.

**Comment by harryzhou2000** [comment]
Hi @lidge-jun — this PR needs a maintainer-sponsored label for the hygiene gate after the latest push: the only file in the auth surface is src/server/auth-cors.ts, where annotateEmptyToolOutputs now gets the same boolean guard the other wire capabilities have (no authentication or credential logic changes). The rebased head also keeps the shared emptiness contract and the image/encrypted/file regressions from your blocker review. Thank you!

**Review by Ingwannu — CHANGES_REQUESTED** [review]
The previous corruption blocker is fixed on eccffb8b7: image, encrypted-content, and file outputs are no longer replaced by the annotation. The existing focused suites pass 58/58 with pinned Bun 1.4.0, and typecheck passes.

One load-bearing Responses bug remains on the current head. The shared helper recognizes only part.type === "text" and "refusal", but real Responses function_call_output arrays use input_text and output_text. I reproduced this through the actual handleResponses path:

- output: [{ type: "input_text", text: "   \n" }] stays as the original array
- output: [{ type: "output_text", text: "\t" }] stays as the original array

Both should become EMPTY_TOOL_OUTPUT_ANNOTATION when the option is enabled. The new test currently uses { type: "text" }, which is the Chat-side internal shape and therefore does not pin the Responses wire contract already covered in tests/responses-parser.test.ts.

Please make the shared predicate wire-aware without widening non-text classification, and add both input_text and output_text regressions through handleResponses while retaining the image/file/encrypted negative cases. Also remove the trailing whitespace at tests/empty-tool-output-annotation.test.ts:100; git diff --check currently fails on the PR head.

After that, rebase the actual branch onto current dev and rerun exact-head CI.

**Comment by harryzhou2000** [comment]
Thanks @Ingwannu — both findings are addressed on the rebased head (57b2b66ad): the shared emptiness predicate is now wire-aware and treats Responses input_text/output_text parts as text-bearing (never widening non-text classification), handleResponses regressions cover whitespace-only input_text and output_text plus a non-empty input_text negative, the image/encrypted/file negatives are retained, and the trailing whitespace at tests/empty-tool-output-annotation.test.ts is gone (git diff --check is clean). Branch rebased onto current dev; focused suites and typecheck pass locally.

File: src/adapters/openai-responses.ts [hunk 1]
```diff
@@ -1705,6 +1736,9 @@ export function createResponsesPassthroughAdapter(provider: OcxProviderConfig):
       if (forward || stateless) {
         outBody = repairOrphanedInputItems(outBody, unexpandedMiss, stateless && !forward);
       }
+      if (provider.annotateEmptyToolOutputs === true) {
+        outBody = annotateEmptyResponsesToolOutputs(outBody, true);
+      }
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply src/adapters/openai-responses.ts]: Addressed on the rebased head (6f15c18d9): annotateEmptyResponsesToolOutputs now runs before repairOrphanedInputItems, so an orphaned empty tool output is annotated first and the repair keeps the annotation text when converting it to a user message. Added two regressions in tests/responses-stateless-dangling-call-repair.test.ts: an empty orphan output carries the [ocx] empty tool output text into its repaired message, and a synthetic missing-result placeholder is never replaced by the annotation. Focused suites (77 tests) and typecheck pass on the exact head.

**Comment by harryzhou2000** [comment]
Fresh self-audit round, pushed as c65423093: missing/null function_call_output.output was previously annotated as “tool ran with no output”, which fabricated execution status for an incomplete payload. It now fails closed (upstream contract applies; orphan repair stays honest). Added regressions for null output and missing output key on the Responses wire, paired + orphaned stateless paths, DeepSeek runtime backfill (saved config without the flag → true; explicit false survives), and the management boolean guard. Focused suites pass (44 + 1), typecheck clean, rebased on dev c44e43f00.

**Comment by harryzhou2000** [comment]
@coderabbitai review

**Comment by harryzhou2000** [comment]
@coderabbitai review

**Comment by harryzhou2000** [comment]
@coderabbitai review

**Comment by harryzhou2000** [comment]
@coderabbitai review

**Comment by harryzhou2000** [comment]
@coderabbitai review

**Comment by harryzhou2000** [comment]
@coderabbitai review

**Comment by harryzhou2000** [comment]
@coderabbitai review

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Review by Ingwannu — APPROVED** [review]
Approved exact head 199df4fd81907ea7d741171ad34fddf0f27e114b. The prior Responses part-type, non-text payload, null and missing output, orphan-order, management validation, and redaction blockers are resolved. Local validation passed 131 focused tests, typecheck, privacy scan, and diff check. All completed exact-head CI jobs are green; the remaining macOS job must finish successfully before merge.

**Comment by Ingwannu** [comment]
Exact-head CI is now fully green, including the macOS full-suite job. I have approved the reviewed head. I am leaving this unmerged because the active dev ruleset requires owner code-owner approval for src/router.ts, which this PR touches; @lidge-jun review remains the final merge gate.

**Comment by harryzhou2000** [comment]
Rebased onto the current upstream dev head (`5dfee1a05`) with no functional changes; the same feature branch was previously approved by Ingwannu at the old head. Local validation on this head: 122 tests pass, typecheck clean. Ready for review.

**Comment by harryzhou2000** [comment]
@Ingwannu PR #2350 was mechanically rebased onto current upstream dev (no functional changes from the head you approved). Gate checks are green on the new head: CodeRabbit pass, enforce-target/hygiene/label pass. The runtime CI workflows still need your approval to run on the new head; could you approve/re-review when convenient?

**Comment by Ingwannu** [comment]
I compared the mechanically rebased series with the previously approved series: all ten feature/fix patches are range-diff equivalent, with only import context moving around the existing Responses changes. I have approved the current exact-head fork workflows.

The new CodeRabbit note about overlapping nested property constraints is a real current-dev gap, but it is outside this PR diff and is not introduced by the empty-tool-output feature. I separated it as #2763 so this PR is not made responsible for an unrelated Moonshot normalizer fix.

This PR remains a draft and still touches `src/router.ts`, so the existing owner/CODEOWNER review gate remains. Please also refresh it onto the latest `dev` before final merge if `dev` advances materially while CI is running.

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@Ingwannu refreshed onto the latest dev (8b1b65b8d) per your note and pushed head 3db0316da. Two CI gates that were red on the old head are unblocked: package.json is bumped to 2.35.0 so the release-version-line guard passes, and the upstream release-train devlog no longer trips the privacy scan (the SSH push line now uses the HTTPS URL — the same repo target, and it applies to the file upstream added after the base I originally rebased onto). Local validation on this head: 122 tests pass, privacy scan clean, typecheck and git diff --check clean. Ready for re-review whenever convenient.

**Comment by harryzhou2000** [comment]
One gate note for @Ingwannu: the package.json bump to 2.35.0 (required by the new release-version-line guard while dev still carries 2.34.0 after the v2.34.0 tag) trips the unsponsored_surface hygiene gate because package.json is a restricted dependency surface. It is the same one-line bump the dev release train will make; if you prefer, bumping dev to 2.35.0 first lets me drop the package.json change from this PR entirely. Otherwise a maintainer-sponsored label would unblock it.

**Comment by harryzhou2000** [comment]
All points from this review are addressed on the current head (e3be23ec4): the option moved to a shared leaf (src/adapters/empty-tool-output-annotation.ts) with one annotation constant and one emptiness contract used by both Chat and Responses; DeepSeek remains the only default-seeded/backfilled provider; image/encrypted/file_id-only Responses outputs are never annotated (regressions added); the management write path now validates the boolean via providerEmptyToolOutputConfigError instead of accepting bad shapes at POST time. The only remaining gate is the package.json version bump (release-version-line), which needs maintainer-sponsored or an upstream dev bump.

**Comment by harryzhou2000** [comment]
Addressed on the current head: the Responses emptiness check is replaced by the shared contract in src/adapters/empty-tool-output-annotation.ts; input_image, encrypted_content, file_id-only and image+whitespace outputs are never annotated (regressions in tests/empty-tool-output-annotation.test.ts all pass); the annotation string is a single shared constant used by both adapters; annotateEmptyToolOutputs is type-checked in the management write path (providerEmptyToolOutputConfigError).

**Comment by Ingwannu** [comment]
Incremental check of current head `e3be23ec45d33a04ddda948ce876cfa441ec4de4`: the new shared text-part contract is correctly used by both Chat and Responses, and it leaves missing/null outputs plus image/file/encrypted-only payloads untouched. The focused adapter regressions passed 44/44 under an isolated HOME on Bun 1.4.0, and `git diff --check` is clean. I do not see a new adapter blocker in this increment. I am not sponsoring the remaining `package.json` bump inside this feature. The canonical post-release base repair is now #2766; after that lands, please rebase onto the resulting `dev`, drop this PR’s package-version commit, and request review again. Exact-head required CI and the existing non-author owner review still need to pass before this leaves Draft.

**Comment by harryzhou2000** [comment]
@Ingwannu #2350 is rebased onto current dev (d1def682d) and the package.json workaround is dropped — upstream's 2.35.0 bump made it unnecessary, so the diff is back to the feature only and hygiene is green (all gates pass, CodeRabbit pass, checklist 4/4, ready). Local 122 tests pass on this head. Could you re-approve/run the fork CI when convenient?

**Comment by harryzhou2000** [comment]
Already done as requested: #2350 is rebased onto current dev (d1def682d, includes #2766) and the package.json change is dropped — the diff is back to the feature only. Gates green on the new head (5422a7728): hygiene/enforce-target/label/resolve-pr pass, checklist 4/4, ready. Local 122 tests pass. Re-requesting your review.

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
The "preserve referenced property constraints during composition" finding is addressed on the current head (081d89bc6): the rebase brought in upstream composeProperties, which intersects overlapping member bounds (MOONSHOT_BOUND_KEYWORDS + intersectBound) and recurses into nested "properties" instead of letting the sibling schema win, while keeping unionRequired and non-overlapping keys untouched. Local adapter suites pass 122/122 on this head with typecheck and diff check clean. CodeRabbit has been resumed on this head.

**Comment by harryzhou2000** [comment]
@Ingwannu #2350 is rebased onto current dev (8d9e28692, includes #2766); head 081d89bc6, feature-only diff, package.json untouched. Gates on this head: hygiene/enforce-target/label/resolve-pr pass, CodeRabbit pass, checklist 4/4, ready. The earlier property-composition finding is resolved by the upstream composeProperties/intersectBound merge carried by this rebase (verified on the head). Local 122 adapter tests pass with typecheck and diff check clean; runtime CI on the fork still needs workflow approval. Ready for your re-review whenever convenient.

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Review by Ingwannu — APPROVED** [review]
Approved exact head 82a9b6f26d80ec4955365a9c6631dd67de499b0e. The ten-commit series is patch-equivalent to the previously reviewed implementation. The current head preserves the Responses input_text/output_text handling, non-text and missing/null safeguards, stateless orphan ordering, management validation, and redacted provider-name error behavior. I reran the advertised exact-head focused suites under isolated HOME: 122 passed, 0 failed, and git diff --check is clean. GitHub reports the head mergeable and all review threads are resolved. Merge only after the repository-required exact-head CI is green.

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@Ingwannu #2350 is refreshed onto current dev (cfb70c972); head 7ca72754f, feature-only diff, CodeRabbit pass, gates green, checklist 4/4, ready. Local adapter suites 122/122 + typecheck clean. Ready for your re-review whenever convenient.

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Review by Ingwannu — CHANGES_REQUESTED** [review]
Review of exact head `bfc71e82ab34082eec665dd258dcc4218d772ab9`: the adapter-side annotation contract is coherent, but the provider overwrite path still has one functional blocker.

`src/server/management/provider-routes.ts:599-603` records ownership of the other fields that registry enrichment can fill, but it does not record whether `annotateEmptyToolOutputs` was submitted. `enrichProviderFromCatalog` then backfills DeepSeek to `true`, and the overwrite at line 645 replaces the stored provider. Therefore an existing DeepSeek provider with an explicit `annotateEmptyToolOutputs: false` is silently changed back to `true` by an unrelated dashboard/provider POST that omits this field. The new PATCH support does not protect this POST path.

Capture `Object.hasOwn(prov, "annotateEmptyToolOutputs")` before enrichment. When it was omitted and the existing provider has a defined boolean, preserve that value before assigning `config.providers[name]`. Add a management regression that starts with DeepSeek set to `false`, performs an unrelated POST overwrite without the field, and asserts the persisted/runtime provider remains `false`.

Keep this draft until that regression passes and the exact-head runtime CI is green. I have not approved or merged this head.

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
@coderabbitai resume

**Comment by harryzhou2000** [comment]
Addressed the review on bfc71e82: provider POST overwrite now captures Object.hasOwn(prov, "annotateEmptyToolOutputs") before registry enrichment and preserves an existing provider's explicit boolean when the payload omits the field, so an unrelated dashboard POST cannot flip an operator's DeepSeek opt-out back to true. Management regression starts DeepSeek at false, performs an unrelated overwrite without the field, and asserts the persisted row stays false. The two CodeRabbit outside-diff findings (retired ollama-cloud model, whitespace-only proxy) are also fixed on this head. Local validation: management-provider-validation 79/79, empty-tool-output + dangling-call + catalog + proxy-env + config 437/437, typecheck and diff-check clean. @coderabbitai resume

**Comment by harryzhou2000** [comment]
Addressed the CodeRabbit findings on 1be941f41: proxy config is now trimmed before environment resolution (whitespace-padded env references resolve correctly) and trimmed again after resolution, and POST /api/providers canonicalizes annotateEmptyToolOutputs null to an omitted field before validation, matching PATCH. Regressions added for both. Head be47bb49f: management-provider-validation + proxy-env 98/98, typecheck and diff-check clean. @coderabbitai resume

File: src/config.ts [hunk 2]
```diff
@@ -3161,7 +3162,7 @@ export function applyProxyEnv(config: OcxConfig): void {
   // malformed values with a privacy-safe warning instead: they cannot express a routing
   // intent, and refusing to start is a worse answer than starting without them.
   const rawProxy = config.proxy;
-  const proxy = typeof rawProxy === "string" ? resolveEnvValue(rawProxy) : undefined;
+  const proxy = typeof rawProxy === "string" ? resolveEnvValue(rawProxy)?.trim() : undefined;
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply src/config.ts]: Fixed on be47bb49f: the raw proxy string is trimmed before resolveEnvValue, so a whitespace-padded env reference resolves correctly, then the resolved value is trimmed again before assignment (whitespace-only values still fall back to direct egress). Regression added for the whitespace-padded reference.

**Comment by harryzhou2000** [comment]
Rebased onto current upstream dev (28c69b03a); head 2f96dc650. Local validation: management-provider-validation 80/80, proxy-env + catalog + config + annotation + dangling-call 438/438, typecheck and diff-check clean. @coderabbitai resume

**Comment by harryzhou2000** [comment]
Addressed the CodeRabbit finding on provider POST null: field presence is now captured before null canonicalization, so a submitted null is an explicit clear and the carry-over guard no longer restores the old override. Regressions: relay null clears an existing true, DeepSeek null clears an existing false and lets registry enrichment re-enable annotation, and the null test now verifies the raw persisted JSON via getConfigPath. Head 76fdde86f: management-provider-validation 82/82, proxy-env + catalog + config + annotation + dangling-call 438/438, typecheck and diff-check clean. @coderabbitai resume

File: tests/management-provider-validation.test.ts [hunk 3]
```diff
@@ -324,6 +325,131 @@ describe("provider management validation", () => {
...
+          name: "relay",
+          provider: {
+            adapter: "openai-chat",
+            baseUrl: "https://relay.example/v1",
+            annotateEmptyToolOutputs: null,
+          },
+        }),
+      });
+      expect(create.status).toBe(200);
+      expect(loadConfig().providers.relay).not.toHaveProperty("annotateEmptyToolOutputs");
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply tests/management-provider-validation.test.ts]: Fixed on 76fdde86f: the null-canonicalization test now first asserts loadConfig().providers.relay is defined, then reads the raw persisted JSON via getConfigPath and verifies the stored relay entry lacks annotateEmptyToolOutputs; a new clear-override test covers an existing true being cleared, and a DeepSeek variant verifies null restores the registry default.

**Comment by harryzhou2000** [comment]
Rebased onto upstream dev 015fedb63; head d996bb02e. Local validation: 348/348 across proxy-env, management-provider-validation, codex-catalog, empty-tool-output annotation; typecheck clean. @coderabbitai resume

**Comment by harryzhou2000** [comment]
Rebased onto upstream dev f4d28626d; head 182d881db. Core suites 318/318 and typecheck clean. @coderabbitai resume

**Comment by lidge-jun** [comment]
Landed on `dev` via #2978 as 8fba9d1b8. All ten of your commits were cherry-picked with authorship preserved, so `git log` credits you rather than the merge — thank you for the feature and for working through the review rounds.

The carry added the fix for the last open blocker (an unrelated provider POST resurrecting the DeepSeek registry default over an explicit `false`), the canonical-OpenAI PATCH overlay fix, and the localized option contract across all seven translated references.

Closing as superseded.
